internal/expr/prepare.go

```diff
@@ -145,54 +123,127 @@ func prepareOutput(ti typeNameToInfo, p *outputPart) (outCols []fullName, typeMe
+	// For example the expressions:
+	//  "(col1, col2) VALUES ($P.*)"
+	//  "(col1, t.col2) AS (&P.*)"
 	if starTypes == 1 && numTypes == 1 {
-		if info, err = fetchInfo(p.targetTypes[0].prefix); err != nil {
+		info, err := ti.lookupInfo(types[0].prefix)
+		if err != nil {
 			return nil, nil, err
 		}
-		for _, c := range p.sourceColumns {
+		for _, c := range columns {
 			tm, err := info.typeMember(c.name)
 			if err != nil {
 				return nil, nil, err
 			}
 			typeMembers = append(typeMembers, tm)
-			outCols = append(outCols, c)
+			genCols = append(genCols, c)
 		}
-		return outCols, typeMembers, nil
+		return genCols, typeMembers, nil
 	} else if starTypes > 0 && numTypes > 1 {
 		return nil, nil, fmt.Errorf("invalid asterisk in types")
 	}
 
-	// Case 3: Explicit columns and types e.g. "(col1, col2) AS (&P.name, &P.id)".
+	// Case 3: Explicit columns and types.
+	// For example the expressions:
+	//  "(col1, col2) VALUES ($P.name, $A.id)"
+	//  "(col1, col2) AS (&P.name, &P.id)"
 	if numColumns == numTypes {
-		for i, c := range p.sourceColumns {
-			t := p.targetTypes[i]
-			if info, err = fetchInfo(t.prefix); err != nil {
+		for i, c := range columns {
+			t := types[i]
+			info, err := ti.lookupInfo(t.prefix)
+			if err != nil {
 				return nil, nil, err
 			}
 			tm, err := info.typeMember(t.name)
 			if err != nil {
 				return nil, nil, err
 			}
 			typeMembers = append(typeMembers, tm)
-			outCols = append(outCols, c)
+			genCols = append(genCols, c)
 		}
 	} else {
 		return nil, nil, fmt.Errorf("mismatched number of columns and target types")
 	}
+	return genCols, typeMembers, nil
+}
 
-	return outCols, typeMembers, nil
+// prepareOutput checks that the output expressions correspond to known types.
+// It then checks they are formatted correctly and finally generates the columns for the query.
+func prepareOutput(ti typeNameToInfo, p *outputPart) ([]fullName, []typeMember, error) {
+	outCols, typeMembers, err := prepareColumnsAndTypes(ti, p.sourceColumns, p.targetTypes)
+	if err != nil {
+		err = fmt.Errorf("output expression: %s: %s", err, p.raw)
+	}
+	return outCols, typeMembers, err
 }
 
-type typeNameToInfo map[string]typeInfo
+// isInsert returns true if the input expression inside an INSERT statement.
+func isInsert(p *inputPart) bool {
+	return len(p.targetColumns) > 0
+}
+
+// prepareInput checks that the input expression is correctly formatted,
+// corresponds to known types, and then generates input columns and values.
+func prepareInput(ti typeNameToInfo, p *inputPart) (inCols []fullName, typeMembers []typeMember, err error) {
+	defer func() {
+		if err != nil {
+			err = fmt.Errorf("input expression: %s: %s", err, p.raw)
+		}
+	}()
+
+	// Check for standalone input expression and prepare if found.
+	// For example:
+	//  "$P.name"
+	if !isInsert(p) {
+		if len(p.sourceTypes) > 1 {
+			return nil, nil, fmt.Errorf("internal error: cannot group standalone input expressions")
```

Comment:
I checked the coverage and this is never executed, either we are missing a test or this statement is unreachable.

Reply:
Ah, I haven't been in the habit of testing internal errors since they would require white box testing which I try to keep to a minimum. In my opinion it is fine to just verify these by inspection, since they are really just a kinder way to crash.

Reply:
Ah so it is just a safeguard for a statement we think is unreachable, then I agree